src/typecheck/core.ml

```diff
@@ -116,9 +117,61 @@ module Ae = struct
 
       (* Semantic triggers *)
       | Type.Builtin (Ast.In_interval (b1, b2)) ->
-        Type.builtin_term (Base.term_app3_ast (module Type) env s
-                 (fun _ast _a1 _a2 _a3 ->
-                    T.in_interval _a1 (b1, b2) _a2 _a3))
+          let recognize_special_bound ast (t : Ast.t) ty =
+            match t.term with
+            | Symbol { name = Simple "?"; _ } -> true, None
+            | Symbol { name = Simple str; _ } ->
+                if String.length str > 0 && String.sub str 0 1 = "?" then
+                  let t =
+                    if ty = Ty.int then
+                      Ast.colon ~loc:t.loc t (Ast.ty_int ~loc:t.loc ())
+                    else if ty = Ty.real then
+                      Ast.colon ~loc:t.loc t (Ast.ty_real ~loc:t.loc ())
+                    else
+                      assert false
+                  in
+                  false, Some t
+                else
+                  false, None
+            | _ -> false, None
+          in
+          let inequality ~strict ~lower =
+            match strict, lower with
+            | true,  true  -> Ast.lt
+            | false, true  -> Ast.leq
+            | true, false  -> Ast.gt
+            | false, false -> Ast.geq
+          in
+          (* The ternary operator in_interval is replaced by two inequalities.meta         We also bind the variables of the form ?id. *)
+          Type.builtin_term (Base.make_op3 (module Type) env s
+            @@ fun ast (t, t1, t2) ->
+              let ty = Type.parse_term env t |> Type.T.ty in
+              let is_omitted1, var1 = recognize_special_bound ast t1 ty in
+              let is_omitted2, var2 = recognize_special_bound ast t2 ty in
+              let loc = t.loc in
+              let res =
+                match is_omitted1, is_omitted2 with
+                | true, true ->
+                    (* This case occurs when the two bounds are infinity. *)
+                    Ast.true_ ~loc ()
+                | true, false ->
+                    inequality ~strict:b2 ~lower:false ~loc t2 t
+                | false, true ->
+                    inequality ~strict:b1 ~lower:true ~loc t1 t
+                | false, false ->
+                    let i1 = inequality ~strict:b1 ~lower:true ~loc t1 t in
+                    let i2 = inequality ~strict:b2 ~lower:false ~loc t2 t in
+                    Ast.and_ ~loc [i1; i2]
+              in
+              let res =
+                match var1 with
+                | Some v -> Ast.exists ~loc [v] res
+                | None -> res
+              in
+              match var2 with
+              | Some v -> Ast.exists [v] res |> Type.parse_term env
```

Comment:
Is this binding the variable at the right level? I think it needs to be at the level of all the triggers (I think the constructor is `Triggers`, not at the level of the `in` — otherwise we forget that we are using the same variable if it occurs in multiple triggers

Reply:
I think you are right. Since there is no explicit bindings in the semantic triggers for these variables, I guess it means the formula is handled as a prenex form and all the quantifiers are at the top level. Nevertheless, we should check if it is the case.

Reply:
The quantifiers must be bound at the level of the closest trigger group (`[ ... ]` brackets).